src/Microsoft.AspNetCore.Authentication.Core/AuthenticationSchemeProvider.cs

```diff
@@ -15,13 +15,28 @@ namespace Microsoft.AspNetCore.Authentication
     public class AuthenticationSchemeProvider : IAuthenticationSchemeProvider
     {
         /// <summary>
-        /// Constructor.
+        /// Creates an instance of <see cref="AuthenticationSchemeProvider"/>
+        /// using the specified <paramref name="options"/>,
         /// </summary>
         /// <param name="options">The <see cref="AuthenticationOptions"/> options.</param>
         public AuthenticationSchemeProvider(IOptions<AuthenticationOptions> options)
+            : this(options, new Dictionary<string, AuthenticationScheme>(StringComparer.Ordinal))
+        {
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="AuthenticationSchemeProvider"/>
+        /// using the specified <paramref name="options"/> and <paramref name="schemes"/>.
+        /// </summary>
+        /// <param name="options">The <see cref="AuthenticationOptions"/> options.</param>
+        /// <param name="schemes">The dictionary used to store authentication schemes.</param>
+        protected AuthenticationSchemeProvider(IOptions<AuthenticationOptions> options, IDictionary<string, AuthenticationScheme> schemes)
         {
             _options = options.Value;
 
+            _schemes = schemes;
```

Comment:
null check

Reply:
Are we allowed to use default parameters?

Since this seems like the cleanest way if so...
```C#
ctor(options, schemes = null) {
  _schemes = schemes ?? new Dictionary()
}
```

Reply:
We actually don't need to have the constructor be protected, since its not like new'ing up one of these does anything without putting it into DI at which point its not like making it protected does anything if they can do that...

Reply:
> Are we allowed to use default parameters?

Default parameters in public methods [is a bad idea](https://haacked.com/archive/2010/08/10/versioning-issues-with-optional-arguments.aspx/)...

Reply:
> We actually don't need to have the constructor be protected

We don't *need to*, but it's much nicer to have just a single public ctor. Besides, why expose more than necessary? Making the ctor public still doesn't help anyone?

Reply:
? This is input from the child class, not DI. It should be checked before using it.

Reply:
I added the check, but didn't make the argument optional or make the ctor public. Should I do that as well?

Reply:
@khellang my point is that no-one is newing these up.  The only reason you would ever call the ctor is if you were extending and calling from base...  I actually am basically agree with your point, a single ctor makes this clearer.  But if we can't use defaults, we have to do an overload, and might as well keep that as protected.  Its only worth having the public/protected debate if we can use default values

Reply:
@Tratcher I don't think its unreasonable for passing in null to mean use the default behavior, its an optional argument after all.

Reply:
But yes it needs to be fixed automatically or throw when its null.

Reply:
Then you would call the other constructor. Don't assign meaning to null.

Reply:
I'm only arguing this if we can get a single public ctor with a default value, otherwise I agree with you guys :)

Reply:
Isn't this type DI activated? Not sure how you'd DI activate the misc dictionary.

Reply:
`services.AddSingleton<IAuthenticationSchemeProvider, YourSchemeProvider>()`

Reply:
The derived ctor news up a dictionary and calls base...

Reply:
Or you insert an instance, we don't care is the point...

Reply:
This actually convinces me more that ctor should be public :)  if its good enough for derived classes, it should be good enough for anyone who wants to new one of these up and stick it in DI as an instance.